packages/processing/pitop/processing/algorithms/ball_detect.py

```diff
@@ -72,88 +72,173 @@ def __circular_match_contour(self):
-        self._radius = value
+class SingleBallDetector:
+    def __init__(self, color: str, image_processing_width: int):
+        self.color = color
+        self.ball = Ball(color)
+        self._image_processing_width = image_processing_width
+        self._frame_scaler = None
 
-    @property
-    def center_points(self) -> deque:
-        """
-        :return: deque of (x, y) coordinates for historical ball detections where (0, 0) is the top left of the frame.
-        :rtype: deque
-        """
-        return self._center_points
+    def draw(self, frame):
+        """Draws the ball on top of the given frame."""
+        frame = ImageFunctions.convert(frame, "OpenCV")
+        self.ball.draw_contrail(frame)
+        if self.found:
+            self.ball.draw_position(frame)
+        return frame
 
-    @property
     def found(self) -> bool:
         """
         :return: Boolean to determine if a valid ball was found in the frame.
         :rtype: bool
         """
-        return self.center is not None
+        return self.ball.valid
 
-    def will_accept(self, ball_likeness: BallLikeness):
-        if ball_likeness.radius < self.MIN_BALL_RADIUS:
+    def will_accept(self, ball: Ball, ball_likeness: BallLikeness):
+        """Check if the ball will accept the new ball likeness based on the
+        radius and shape accuracy."""
+        if ball_likeness.radius < ball.MIN_RADIUS:
             return False
 
-        if ball_likeness.circular_likeness < self.minimum_shape_accuracy:
+        if ball_likeness.circular_likeness < ball.minimum_shape_accuracy:
             return True
 
-        if ball_likeness.radius > self.BALL_CLOSE_RADIUS:
+        if ball_likeness.radius > ball.CLOSE_RADIUS:
             return True
 
         return False
 
+    def _do_find(self, frame) -> Ball:
+        """Finds a ball in a frame."""
+        contours = find_contours(frame, self.color)
+        ball = self.ball
+
+        # Reset ball state
+        ball.center = None
+        ball.radius = 0
+        ball.angle = None
+
+        if len(contours) == 0:
+            # No ball found
+            ball.center_points.appendleft(None)
+            return ball
+
+        highest_likelihood = 0
+
+        # Check each contour for ball likeness
+        for contour in contours:
+            ball_likeness = BallLikeness(contour)
+
+            if ball_likeness.likelihood < highest_likelihood:
+                continue
+
+            # Don't accept future balls which are less likely
+            highest_likelihood = ball_likeness.likelihood
+            if self.will_accept(ball, ball_likeness):
+                ball.center = ball_likeness.pos
+                ball.radius = ball_likeness.radius
+
+        return ball
+
+    def find(self, frame) -> Ball:
+        """Finds a ball in a frame.
+
+        Scales down the provided frame to reduce processing time.
+        """
+
+        # Scale image
+        if self._frame_scaler is None:
```

Comment:
This can probably be removed by creating a new frame_scaler each frame (right now passing varying frame sizes will be slightly broken by the incorrect _frame_scaler value being used to correct the coordinates)